### PR TITLE
Integration Back End - Front End

### PR DESCRIPTION
**Pull request by nicotom**

### CODE REVIEW

File: __init__.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+from webpage.webstreaming import *
```
Comment (nicotom) [inline __init__.py]: `from webpage.webstreaming import start_webstreaming`

Comment (nicotom) [reply __init__.py]: I would create a new file for this

File: detector3.py [hunk 2]
```diff
@@ -46,7 +46,7 @@ def __init__(self, ip, port=1865):
         print('Max integration time:', self.MAX_INTEGRATION_TIME)
         
         self.MAX_INTENSITY        = int(self._spectrometer.get_max_intensity())
-        print('Max intensity:', self.MAX_INTENSITY)
+        print('Max intensity:', self.MAX_INTENSITY,'\n')
```
Comment (nicotom) [inline detector3.py:48]: This seems unnecesary

File: webpage/webstreaming.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+import argparse
+from flask import Flask, url_for, jsonify
+from flask import render_template
+from flask_bootstrap import Bootstrap
+from flask import request, redirect
+#from flask_appconfig import AppConfig
+from webpage.config import *
+from webpage.forms import ConfigForm
+from time import sleep
+from detector3 import Detector
```
Comment (nicotom) [inline webpage/webstreaming.py]: This uses only detector3 instead of detector, we should remove the py2 only version

File: detector3.py [hunk 4]
```diff
@@ -135,7 +152,7 @@ def _save_spectrum(self, path, spectrum):
             Wavelengths Intensities
         '''
         f = os.path.join(path, '%s.txt' % datetime.strftime(datetime.now(), '%d-%m-%Y_%H:%M:%S'))
-        print("Saving in '%s'" % f)
+        print("Saving in '%s'\n" % f)
```
Comment (nicotom) [inline detector3.py:153]: also unnecesary

File: detector3.py [hunk 5]
```diff
@@ -179,47 +196,83 @@ def run(self):
...
+    def button_start(self):
+        print("GPIO START event")
+        self._gpio_started = True
+
+    def button_stop(self):
+        print("GPIO STOP event")
+        self._gpio_started = False
+
+    def configure_gpio(self):
+        import RPi.GPIO as GPIO
```
Comment (nicotom) [inline detector3.py]: The imports should be at the top of the file

File: exp/outputs/load_data.py [hunk 6]
```diff
@@ -0,0 +1,16420 @@
...
+302.39,
+322.73,
+245.9,
+293.35,
+356.63,
+320.47,
+315.95,
+277.54,
+277.54,
+277.54]
```
Comment (nicotom) [inline exp/outputs/load_data.py]: This must be a file to read not a list in python, it makes the code unreadable.

File: spectrometer3.py [hunk 7]
```diff
@@ -23,7 +23,7 @@
 
 import socket
 from sys import exit
-#from time import sleep
+from time import sleep#,time
```
Comment (nicotom) [inline spectrometer3.py]: no change

Comment (nicotom) [reply spectrometer3.py]: To mock we should inherit from Spectrometer class and change only the mocked methods

File: webpage/webstreaming.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+                __name__,
+                template_folder="templates",
+                static_folder="static"
+                )
+
+    Bootstrap(app)
+
+    app.config['SECRET_KEY'] = 'devkey'
+    app.config['RECAPTCHA_PUBLIC_KEY'] = \
+        '6Lfol9cSAAAAADAkodaYl9wvQCwBMr3qGR_PPHcw'
```
Comment (nicotom) [inline webpage/webstreaming.py]: showing the public key?

File: webpage/webstreaming.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        integration_time                    = get_spectro_scope('integration_time',app)*1000 #Pass integration time to ms
+        det.integration_time    = get_spectro_scope('integration_time',app)
+        det.integration_factor  = get_spectro_scope('integration_factor',app)
+        det.threshold           = get_spectro_scope('threshold',app)
+        return render_template("spectroscope.html",data_x=det.get_wavelengths(),integration_time=integration_time,form=form, auto_en= (True if "automatic" == det.operation_mode else False))
+        #return render_template("spectroscope.html",data_x=det.get_wavelengths(),integration_time=int(det.integration_time*1000),form=form)
+
+    @app.route("/default",methods=['GET','POST'])
+    def set_default_config():
+        if request.method == 'POST':
```
Comment (nicotom) [inline webpage/webstreaming.py]: why this if?

File: webpage/webstreaming.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    ap = argparse.ArgumentParser()
+
+    ap.add_argument("-i", "--ip", type=str, required=True,
+        help="ip address of the device")
+
+    ap.add_argument("-o", "--port", type=int, required=True,
+        help="ephemeral port number of the server (1024 to 65535)")
+
+    ap.add_argument("-loc", "--location", type=str, required=True,
+        help="ephemeral port number of the server (1024 to 65535)")
```
Comment (nicotom) [inline webpage/webstreaming.py]: wrong help

File: webpage/webstreaming.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+    ap.add_argument("-id","--ip_det", type=str, required=True,
+        help="ip address of the device")
```
Comment (nicotom) [inline webpage/webstreaming.py]: what does this argument mean?

File: webpage/webstreaming.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
+from webpage.utils import *
+#from tracker_lib import *
```
Comment (nicotom) [inline webpage/webstreaming.py]: Remove this commented code

File: __init__.py (hunk 1 above)
Comment (facu-s) [reply __init__.py]: Agree

File: detector3.py (hunk 2 above)
Comment (facu-s) [reply detector3.py:48]: Agree

File: webpage/webstreaming.py (hunk 8 above)
Comment (facu-s) [reply webpage/webstreaming.py]: This is exactly the same as the previous project. I think there's no need to hide it right now.

File: webpage/webstreaming.py (hunk 9 above)
Comment (facu-s) [reply webpage/webstreaming.py]: This function is still deprecated for this project. We need to modify it once we have something functional.

File: webpage/webstreaming.py (hunk 10 above)
Comment (facu-s) [reply webpage/webstreaming.py]: You are right. That was a fast copy&paste and i forgot to change the description.

File: webpage/webstreaming.py (hunk 11 above)
Comment (facu-s) [reply webpage/webstreaming.py]: Same comment above.

**Review by facu-s — APPROVED** [review]
I downloaded the last changed on the branch. The only thing i had a problem with was with the installation of pandas on the Raspberry, but was an internal error. I solved it by creating a venv, and i had to add RPi.GPIO to the requirements as well.

After that minor change, the rest works like a charm.

File: webpage/webstreaming.py (hunk 10 above)
Comment (nicotom) [reply webpage/webstreaming.py]: I changed this help, so it's now OK

File: webpage/webstreaming.py (hunk 11 above)
Comment (nicotom) [reply webpage/webstreaming.py]: I don't know what to put in help, can you change it @facu-s ?

**Comment by nicotom** [comment]
@facu-s can you add RPi.GPIO in the requirements file?

File: webpage/webstreaming.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+    spectrometer = MockSpectrometer(ip_address=args["ip_det"], port=1865, channel=0)
+    det = MockDetector(spectrometer)
+    app = create_app(det)
+    init_db(app)
+
+    det.location = args["location"]
+    det.configure_gpio()
+    det.start()
+
+    app.run(host=args["ip"], port=args["port"], debug=True, threaded=True, use_reloader=False)
```
Comment (nicotom) [inline webpage/webstreaming.py:107]: @facu-s do you think that we can move this file to root? just like the other project?

Comment (facu-s) [reply webpage/webstreaming.py:107]: The principal idea to leave that file inside a specific folder was to have an encapsulated environment for the webpage, thinking on future projects. However I like the idea on having a file at root, working as an interface with the webpage environment.

From my point of view if we are willing to make a change, then we should split the creation of the app of Flask and the definition of it in two different files. Where one should be located inside the webpage folder (the one containing the definition) and the other at root as you say.

File: webpage/webstreaming.py (hunk 11 above)
Comment (facu-s) [reply webpage/webstreaming.py]: yes

**Pull request merged**